core/src/main/java/com/linecorp/armeria/common/HttpRequest.java

```diff
@@ -280,6 +282,55 @@ static HttpRequest of(RequestHeaders headers, Publisher<? extends HttpObject> pu
         }
     }
 
+    /**
+     * Creates a new HTTP request whose {@link Publisher} is produced by the specified
+     * {@link CompletionStage}. If the specified {@link CompletionStage} fails, the returned request will be
+     * closed with the same cause as well.
+     *
+     * @param stage the {@link CompletionStage} which will produce the actual data and trailers
+     */
+    static HttpRequest from(RequestHeaders headers,
+                            CompletionStage<? extends StreamMessage<? extends HttpObject>> stage) {
+        requireNonNull(headers, "headers");
+        requireNonNull(stage, "stage");
+
+        return of(headers, StreamMessage.of(stage));
+    }
+
+    /**
+     * Creates a new HTTP request whose {@link Publisher} is produced by the specified
+     * {@link CompletableFuture}. If the specified {@link CompletableFuture} fails, the returned request
+     * will be closed with the same cause as well.
+     *
+     * @param future the {@link CompletableFuture} which will produce the actual data and trailers
+     */
+    static HttpRequest from(RequestHeaders headers,
+                            CompletableFuture<? extends StreamMessage<? extends HttpObject>> future) {
```

Comment:
Question) Is this API needed since `CompletionFuture` inherits `CompletionStage`?

Reply:
Oops, will remove it.